### PR TITLE
`Integrated code lifecycle`: Add option to clone repositories using SSH

### PR DESCRIPTION
**Pull request by krusche**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

The process of authenticating to repositories from localVC, it involves several steps. Using SSH as authentication simplifies this.

### Description

Added ssh authentication for means to clone repositories from localVC. Also added a rudimentary Ssh Client for testing the functionality. Disclaimer: so far this test only initializes the ssh Server in Artemis, and an Ssh client that sends "git push" requests over ssh. More test functionality will be added in a future PR.

### Steps for Testing

1. In the user account settings, add a public ssh key (locally generated with e.g. `ssh-keygen -t rsa`
2. As an instructor create a programming exercise.
3. View the exercise's details page, and copy the ssh URL by using the 'clone repository' button
4. Now locally, clone the repository: `git clone <copied link>`

#### Exam Mode Testing

It should behave identical in programming exercises in the exam mode.

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

### Screenshots

SSH clone option in the client now with ICL 🚀 
image

Edit in User Settings:
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java [hunk 1]
```diff
@@ -194,4 +196,19 @@ public ResponseEntity<ZonedDateTime> getIrisAcceptedForStudent() {
...
+    /**
+     * PUT users/sshpublickey : sets the ssh public key
+     *
+     * @return the ResponseEntity with status 200 (OK), with status 404 (Not Found), or with status 400 (Bad Request)
+     */
+    @PutMapping("users/sshpublickey")
+    @EnforceAtLeastStudent
+    public ResponseEntity<Void> addSshPublicKey(String sshPublicKey) throws NoSuchAlgorithmException {
+        User user = userRepository.getUser();
+        var sshPublicKeyHash = HashUtils.hashString(sshPublicKey);
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:209]: We should not hash the public key

**Review by bassner — CHANGES_REQUESTED** [review]

**Review by Hialus — CHANGES_REQUESTED** [review]
Me and @bensofficial tried to test this locally and were sadly unable to successfully authenticate using our SSH keys. We tried to find a fix, but were unable to for now.

I also found some potential code improvements while doing so.

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java [hunk 2]
```diff
@@ -194,4 +196,19 @@ public ResponseEntity<ZonedDateTime> getIrisAcceptedForStudent() {
...
     }
+
+    /**
+     * PUT users/sshpublickey : sets the ssh public key
+     *
+     * @return the ResponseEntity with status 200 (OK), with status 404 (Not Found), or with status 400 (Bad Request)
+     */
+    @PutMapping("users/sshpublickey")
+    @EnforceAtLeastStudent
+    public ResponseEntity<Void> addSshPublicKey(String sshPublicKey) throws NoSuchAlgorithmException {
```
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:207]: This endpoint won't work without `@RequestBody`.
```suggestion
    public ResponseEntity<Void> addSshPublicKey(@RequestBody @NotNull String sshPublicKey) throws NoSuchAlgorithmException {
```

File: docker/nginx/nginx.conf [hunk 3]
```diff
@@ -39,3 +39,13 @@ http {
...
+
+stream {
+    upstream artemis {
+        include includes/artemis-ssh-upstream.conf;
+    }
+    server {
+        listen 7921;
+        proxy_pass ssh;
+    }
+}
```
[the remark this answers is not in the document]
Comment (Hialus) [reply docker/nginx/nginx.conf]: Fixed

File: src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/SshConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+@Configuration
+public class SshConfiguration {
+
+    private static final Logger log = LoggerFactory.getLogger(SshConfiguration.class);
+
+    @Value("${artemis.version-control.ssh-port:7921}")
+    private int sshPort;
+
+    @Value("${artemis.version-control.ssh-host-key-path:null}")
+    private Optional<Path> sshHostKeyPath;
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/SshConfiguration.java:34]: Would be fine for me. However, I don't know if this is supported by the Apache mina sshd framework. We need to double check

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java (hunk 1 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java]: Hashing the key is needed for fast lookup in the database. We now store the hashed value for fast lookup with a database index and the full key to display it in the client. This should be a good compromise

File: src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/SshConfiguration.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/SshConfiguration.java:31]: I've now managed to add support for this in [ab48db5](https://github.com/ls1intum/Artemis/pull/8110/commits/ab48db56193a96fe293ec16b4c53635fb55bc9ea)

File: src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/MultipleHostKeyProvider.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                    }
+                    catch (Exception e) {
+                        warn("resolveKeyPair({}) Failed ({}) to load: {}", file, e.getClass().getSimpleName(), e.getMessage(), e);
+                    }
+                }
+            });
+        }
+        catch (IOException e) {
+            warn("loadKeys({}) Failed ({}) to list keys: {}", path, e.getClass().getSimpleName(), e.getMessage(), e);
+        }
```
[the remark this answers is not in the document]
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/MultipleHostKeyProvider.java:66]: Not needed here. If the host key loading fails, the SSH connection will fail, which is error handling enough. We simply need to have a log statement here to know what went wrong if something went wrong

**Review by kaancayli — APPROVED** [review]
Tested on ts1. Works as expected. Code lgtm

**Review by az108 — DISMISSED** [review]
tested on TS1 and works as expected without any errors

**Review by az108 — APPROVED** [review]
LGTM and tested again

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCService.java [hunk 6]
```diff
@@ -56,8 +56,12 @@ public class LocalVCService extends AbstractVersionControlService {
     @Value("${artemis.version-control.url}")
     private URL localVCBaseUrl;
 
+    private static String localVCBasePath;
```
Comment (magaupp) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCService.java:59]: Why are these fields static? The fields and their usages could be made non-static simplifying injection.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCServletService.java [hunk 7]
```diff
@@ -88,10 +91,12 @@ public class LocalVCServletService {
 
     private final ProgrammingTriggerService programmingTriggerService;
 
-    private final LocalVCService localVCService;
+    private static URL localVCBaseUrl;
```
Comment (magaupp) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCServletService.java:94]: See [my other comment](https://github.com/ls1intum/Artemis/pull/8110/files#r1605804943).

**Review by Jan-Thurner — APPROVED** [review]
Cloning exercises in the exam mode and normal exercises tested on TS1. Works as expected

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCService.java (hunk 6 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCService.java:59]: If I understand correctly, we want to make sure to keep the localVCBasePath the same across all LocalVCService instances. Though in this case it probably would not change anything removing the static, as there is only one instance of the LocalVCService (if I'm not mistaken)

Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCService.java:59]: We need to access the variable without having an instance of the LocalVCService, in e.g. line 62

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCServletService.java (hunk 7 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCServletService.java:94]: same as above

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCServletService.java [hunk 8]
```diff
@@ -88,10 +91,12 @@ public class LocalVCServletService {
...
 
     @Value("${artemis.version-control.url}")
-    private URL localVCBaseUrl;
+    public void setLocalVCBaseUrl(URL localVCBaseUrl) {
+        LocalVCServletService.localVCBaseUrl = localVCBaseUrl;
+    }
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCServletService.java:99]: In this case we need it to be static to access it without instantiating an object first.

File: src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCSSHTest.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        SshServer sshServer = sshConfiguration.sshServer();
+
+        localVCLocalCITestService.createParticipation(programmingExercise, student1Login);
+
+        // this command arrives at the ssh server if you would manually push from the command line
+        String commandString = "git-receive-pack '/git/" + projectKey1 + "/" + templateRepositorySlug + "'";
+        SshGitCommandFactoryService sfd = (SshGitCommandFactoryService) sshServer.getCommandFactory();
+        SshGitCommand command = (SshGitCommand) sfd.createGitCommand(commandString);
+        // command.setSession(....);
+        // TODO set serverSession in command and authenticate
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCSSHTest.java]: I could not yet figure out how to properly retrieve, and then set the ServerSession, if you have any ideas how to do this feel free to help.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java [hunk 10]
```diff
@@ -231,7 +229,7 @@ private void checkCorrectBranchElseThrow(final ProgrammingExerciseParticipation
             if (participation instanceof ProgrammingExerciseStudentParticipation studentParticipation) {
                 participationDefaultBranch = versionControlService.orElseThrow().getOrRetrieveBranchOfStudentParticipation(studentParticipation);
             }
-            if (Strings.isNullOrEmpty(participationDefaultBranch)) {
+            if (StringUtils.isEmpty(participationDefaultBranch)) {
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:232]: This is unrelated to this PR

File: src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/service/SshGitLocationResolverService.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        var repositoryAction = gitCommand.equals("git-upload-pack") ? RepositoryActionType.READ : gitCommand.equals("git-receive-pack") ? RepositoryActionType.WRITE : null;
+        try {
+            localVCServletService.authorizeUser(repositoryTypeOrUserName, user, exercise, repositoryAction, localVCRepositoryUri.isPracticeRepository());
+        }
+        catch (LocalVCForbiddenException e) {
+            log.error("User {} does not have access to the repository {}", user.getLogin(), repositoryPath);
+            throw new AccessDeniedException("User does not have access to this repository", e);
+        }
+
+        try (Repository repo = localVCServletService.resolveRepository(repositoryPath)) {
```
Comment (magaupp) [inline src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/service/SshGitLocationResolverService.java:84]:
```suggestion
        // we cannot trust unvalidated user input
        String localRepositoryPath = localVCRepositoryUri.getRelativeRepositoryPath().toString();
        try (Repository repo = localVCServletService.resolveRepository(localRepositoryPath)) {
```

**Review by magaupp — CHANGES_REQUESTED** [review]
The handling of the user-supplied path is concerning. Students are able to supply `%2e%2e/` (`../`) in the path and navigate to other repositories. If a student was able to create directories with matching slug and project key in a different repository, they might be able to clone that repository.

I suggest to reconstruct the expected path from the validated parts.

I would add this to `LocalVCRepositoryUri` (I don't know how to add comments to files without changes):
```patch
Index: src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCRepositoryUri.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCRepositoryUri.java b/src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCRepositoryUri.java
--- a/src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCRepositoryUri.java	(revision 3da6196bef4a793917ccd1574d8beb0e1048d585)
+++ b/src/main/java/de/tum/in/www1/artemis/service/connectors/localvc/LocalVCRepositoryUri.java	(date 1716476335288)
@@ -278,6 +278,19 @@
      * @return The full Path to the repository, which includes the base path, project key, and repository slug with a ".git" suffix.
      */
     public Path getLocalRepositoryPath(String localVCBasePath) {
-        return Paths.get(localVCBasePath, projectKey, repositorySlug + ".git");
+        Path relativeRepositoryPath = getRelativeRepositoryPath();
+        return Path.of(localVCBasePath).resolve(relativeRepositoryPath);
+    }
+
+    /**
+     * Computes and returns the path to the repository stored within the local version control system relative to the base path.
+     * This path is constructed using the project key and repository slug.
+     * The result is a Path object that can be resolved against the base path
+     * to get the directory where the repository is stored or should be stored locally.
+     *
+     * @return The relative Path to the repository, which includes the project key and repository slug with a ".git" suffix.
+     */
+    public Path getRelativeRepositoryPath() {
+        return Paths.get(projectKey, repositorySlug + ".git");
     }
 }
```

File: src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/service/SshGitLocationResolverService.java (hunk 11 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/service/SshGitLocationResolverService.java]: That's actually a good point. I added your suggestion

File: src/test/java/de/tum/in/www1/artemis/localvcci/MultipleHostKeyProviderTest.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.nio.file.Path;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.context.annotation.Profile;
+
+import de.tum.in.www1.artemis.config.localvcci.ssh.MultipleHostKeyProvider;
+
+@Profile(PROFILE_LOCALVC)
+public class MultipleHostKeyProviderTest extends LocalVCIntegrationTest {
```
Comment (SimonEntholzer) [inline src/test/java/de/tum/in/www1/artemis/localvcci/MultipleHostKeyProviderTest.java:14]:
```suggestion
class MultipleHostKeyProviderTest extends LocalVCIntegrationTest {
```

**Review by Jan-Thurner — DISMISSED** [review]
Tested on TS1.
  Cloning and pushing regular programming exercises with SSH: ✅ 
  Cloning and pushing exam mode programming exercises with SSH: ✅

File: src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCSshIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        }
+
+        commandString = "git-receive-pack '/git/" + projectKey1 + "/" + templateRepositorySlug + "'";
+        try {
+            setupCommand(commandString, (ServerSession) serverSession).run();
+        }
+        catch (NullPointerException e) {
+            assertThat(e).isInstanceOf(NullPointerException.class);
+            assertThat(false).isTrue();
+        }
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCSshIntegrationTest.java:115]: Here we fail the test when we encounter such an Exception.

File: src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCSshIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+        // this command arrives at the ssh server if you would manually push from the command line
+        String commandString = "git-receive-pack '/git/" + projectKey1 + "/" + templateRepositorySlug + "'";
+        SshGitCommandFactoryService sshGitCommandFactory = (SshGitCommandFactoryService) sshServer.getCommandFactory();
+        SshGitCommand command = (SshGitCommand) sshGitCommandFactory.createGitCommand(commandString);
+        try {
+            command.run();
+        }
+        catch (NullPointerException e) {
+            assertThat(e).isInstanceOf(NullPointerException.class);
+        }
```
[the remark this answers is not in the document]
Comment (SimonEntholzer) [reply src/test/java/de/tum/in/www1/artemis/localvcci/LocalVCSshIntegrationTest.java:63]: This is true, we will improve this test in a future PR

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java [hunk 15]
```diff
@@ -19,7 +19,7 @@ public final class Constants {
 
     public static final int PASSWORD_MIN_LENGTH = 8;
 
-    public static final int PASSWORD_MAX_LENGTH = 50;
+    public static final int PASSWORD_MAX_LENGTH = 100;
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/config/Constants.java:22]: Was this intentionally changed in this PR?

File: src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/SshConfiguration.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+            // this allows to customize the host key file for production environments
+            log.info("Use SSH host key {}", sshHostKeyPath.get());
+            sshd.setKeyPairProvider(new MultipleHostKeyProvider(sshHostKeyPath.get()));
+        }
+        else {
+            // this is a simple solution for development, the host key will be generated during the first ssh operation in case it does not exist
+            sshd.setKeyPairProvider(new SimpleGeneratorHostKeyProvider(Paths.get("tmp", "hostkey.ser")));
+        }
+        sshd.setCommandFactory(
+                sshGitCommandFactoryService.withGitLocationResolver(sshGitLocationResolverService).withExecutorServiceProvider(() -> ThreadUtils.newFixedThreadPool("git-ssh", 8)));
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/SshConfiguration.java:73]: Why do we use a fixed thread pool of exactly size 8? Should we use a CachedThreadPool to make this more dynamic?

File: src/test/javascript/spec/component/account/account-information.component.spec.ts [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    });
+
+    beforeEach(() => {
+        profileServiceMock.getProfileInfo.mockReturnValue(of({ activeProfiles: [PROFILE_LOCALVC] }));
+        accountServiceMock.getAuthenticationState.mockReturnValue(of({ id: 99, sshPublicKey: mockKey } as User));
+    });
+
+    it('should create', () => {
+        expect(comp).toBeTruthy();
+    });
```
Comment (bassner) [inline src/test/javascript/spec/component/account/account-information.component.spec.ts:55]:
```suggestion
```

**Review by bassner — CHANGES_REQUESTED** [review]
Tested the new feature on TS1 - works without problems 👍🏽 

I recommend to improve the UX in a follow up PR so ...
- the UI selects the SSH clone URL as default if the user has a public key set up
- display a hint that Artemis supports SSH in the clone popup
- add a way to remove your public key in the settings
- add a more informative error if you make a mistake adding it; right now it's only "Bad Request" or "Internal Server Error".

The code looks good. I added some minor comments.
However, some files are completely undocumented, especially in `src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/**/*`. It's hard to understand what's happening in some parts of it. I recommend adding class-level and method-level javadoc and inline comments for the more complex parts to improve clarity for developers who are not deeply involved in ICL

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java (hunk 15 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/config/Constants.java:22]: I would assume it was, but I'm not sure tbh @Hialus

File: src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/SshConfiguration.java (hunk 16 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/SshConfiguration.java:73]: Tbh I don't know. I'll keep this in mind and try it out in a follow-up PR

**Comment by SimonEntholzer** [comment]
> Tested the new feature on TS1 - works without problems 👍🏽
> 
> I recommend to improve the UX in a follow up PR so ...
> 
>     * the UI selects the SSH clone URL as default if the user has a public key set up
> 
>     * display a hint that Artemis supports SSH in the clone popup
> 
>     * add a way to remove your public key in the settings
> 
>     * add a more informative error if you make a mistake adding it; right now it's only "Bad Request" or "Internal Server Error".
> 
> 
> The code looks good. I added some minor comments. However, some files are completely undocumented, especially in `src/main/java/de/tum/in/www1/artemis/config/localvcci/ssh/**/*`. It's hard to understand what's happening in some parts of it. I recommend adding class-level and method-level javadoc and inline comments for the more complex parts to improve clarity for developers who are not deeply involved in ICL

Thanks for the UI suggestion, I'll add them in the follow-up PR. :+1:

**Review by kaancayli — CHANGES_REQUESTED** [review]
Tested on TS2. Could clone the repo without issues. I have left a small comment.

Other than that I think it might be helpful to give the user a hint if they saved their ssh key or not. I think ssh option should not be visible, when user did not save their ssh key.

**Review by kaancayli — APPROVED** [review]

**Pull request merged**